src/planner/Dijkstra.cpp

```diff
@@ -0,0 +1,158 @@
+//        this list of conditions and the following disclaimer.
+//     2. Redistributions in binary form must reproduce the above copyright
+//        notice, this list of conditions and the following disclaimer in the
+//        documentation and/or other materials provided with the distribution.
+//     3. Neither the name of the copyright holder nor the names of its
+//        contributors may be used to endorse or promote products derived from
+//        this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+////////////////////////////////////////////////////////////////////////////////
+
+#include "planner/Dijkstra.hpp"
+#include <iostream>
+
+namespace libcozmo {
+namespace planner {
+
+    bool Dijkstra::set_start(const int& start_id) {
+        if (m_state_space->is_valid_state(
+                *m_state_space->get_state(start_id))) {
+            m_start_id = start_id;
+            return true;
+        }
+        return false;
+    }
+
+    bool Dijkstra::set_goal(const int& goal_id) {
+        if (m_state_space->is_valid_state(
+                *m_state_space->get_state(goal_id))) {
+            m_goal_id = goal_id;
+            return true;
+        }
+        return false;
+    }
+
+    bool Dijkstra::solve(std::vector<int>* actions) {
+        if (m_goal_id == -1 || m_start_id == -1) {
+            return false;
+        }
+        CostMap costmap;
+        ChildToParentMap child_to_parent_map;
+        CostMapComparator comparator(costmap);
+        std::set<int, CostMapComparator> Q(comparator);
+        std::vector<int> path_ids;
+        Q.insert(m_start_id);
+        costmap[m_start_id] = 0;
+        while (!Q.empty()) {
+            int curr_state_id = *(Q.begin());
+            Q.erase(Q.begin());
+            Q.erase(curr_state_id);
+            if (is_goal(curr_state_id)) {
+                extract_sequence(
+                    child_to_parent_map, m_start_id, m_goal_id, actions);
+                return true;
+            }
+            const statespace::StateSpace::State* curr_state =
+                m_state_space->get_state(curr_state_id);
+            for (int i = 0; i < m_action_space->size(); i++) {
```

Comment:
The loop through the actions should be in a separate function called `get_successors`